File: generated_code/cpp/test_unittest_671.cpp
#include <gtest/gtest.h>
#include <vector>
#include <queue>
#include "solution_proxy.h"

// Helper function to build a tree from a level-order vector (using -1 for null/empty nodes as per constraint Node.val >= 1)
TreeNode* buildTree(const std::vector<int>& nodes) {
    if (nodes.empty() || nodes[0] == -1) return nullptr;
    
    TreeNode* root = new TreeNode(nodes[0]);
    std::queue<TreeNode*> q;
    q.push(root);
    
    size_t i = 1;
    while (!q.empty() && i < nodes.size()) {
        TreeNode* curr = q.front();
        q.pop();
        
        if (i < nodes.size() && nodes[i] != -1) {
            curr->left = new TreeNode(nodes[i]);
            q.push(curr->left);
        }
        i++;
        
        if (i < nodes.size() && nodes[i] != -1) {
            curr->right = new TreeNode(nodes[i]);
            q.push(curr->right);
        }
        i++;
    }
    return root;
}

// Helper function to free the allocated tree memory
void freeTree(TreeNode* root) {
    if (!root) return;
    freeTree(root->left);
    freeTree(root->right);
    delete root;
}

class SecondMinimumValueTest : public ::testing::Test {
protected:
    Solution solution;
};

// Test Case 1: Provided Example 1
// Tree: [2, 2, 5, null, null, 5, 7]
// Expected: 5
TEST_F(SecondMinimumValueTest, Example1_StandardCase) {
    std::vector<int> nodes = {2, 2, 5, -1, -1, 5, 7};
    TreeNode* root = buildTree(nodes);
    
    EXPECT_EQ(solution.findSecondMinimumValue(root), 5);
    
    freeTree(root);
}

// Test Case 2: Provided Example 2
// Tree: [2, 2, 2]
// Expected: -1 (No second minimum exists)
TEST_F(SecondMinimumValueTest, Example2_AllNodesEqual) {
    std::vector<int> nodes = {2, 2, 2};
    TreeNode* root = buildTree(nodes);
    
    EXPECT_EQ(solution.findSecondMinimumValue(root), -1);
    
    freeTree(root);
}

// Test Case 3: Minimum size constraint (Single node tree)
// Tree: [5]
// Expected: -1
TEST_F(SecondMinimumValueTest, EdgeCase_SingleNodeTree) {
    std::vector<int> nodes = {5};
    TreeNode* root = buildTree(nodes);
    
    EXPECT_EQ(solution.findSecondMinimumValue(root), -1);
    
    freeTree(root);
}

// Test Case 4: Values at maximum boundary (2^31 - 1)
// Tree: [2, 2, 2147483647]
// Expected: 2147483647
TEST_F(SecondMinimumValueTest, EdgeCase_MaxIntValue) {
    std::vector<int> nodes = {2, 2, 2147483647};
    TreeNode* root = buildTree(nodes);
    
    EXPECT_EQ(solution.findSecondMinimumValue(root), 2147483647);
    
    freeTree(root);
}

// Test Case 5: Second minimum is on the left subtree instead of right
// Tree: [3, 4, 3, 4, 5]
// Expected: 4
TEST_F(SecondMinimumValueTest, Scenario_SecondMinimumOnLeft) {
    std::vector<int> nodes = {3, 4, 3, 4, 5};
    TreeNode* root = buildTree(nodes);
    
    EXPECT_EQ(solution.findSecondMinimumValue(root), 4);
    
    freeTree(root);
}

// Test Case 6: Deep skewed tree property holding root.val == min(left, right)
// Tree: [1, 1, 3, 1, 2]
// Expected: 2
TEST_F(SecondMinimumValueTest, Scenario_DeepTreeMultipleValues) {
    std::vector<int> nodes = {1, 1, 3, 1, 2};
    TreeNode* root = buildTree(nodes);
    
    EXPECT_EQ(solution.findSecondMinimumValue(root), 2);
    
    freeTree(root);
}

File: generated_code/python/test_unittest_671.py
import os
import importlib.util
import unittest
from typing import Optional

# Dynamic loading of the solution module as per guidelines
solution_path = os.environ.get("TEST_SOLUTION_FILE")
if not solution_path:
    raise RuntimeError("TEST_SOLUTION_FILE environment variable is not set.")

spec = importlib.util.spec_from_file_location("Solution", solution_path)
sol_module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(sol_module)
Solution = sol_module.Solution

# Retrieve TreeNode from the loaded module or define a compatible one if not exposed
if hasattr(sol_module, 'TreeNode'):
    TreeNode = sol_module.TreeNode
else:
    class TreeNode:
        def __init__(self, val=0, left=None, right=None):
            self.val = val
            self.left = left
            self.right = right

class TestFindSecondMinimumValue(unittest.TestCase):

    def test_example_1_has_second_minimum(self):
        """Test Example 1: Tree with distinct values where second minimum exists."""
        # Tree structure: [2, 2, 5, null, null, 5, 7]
        root = TreeNode(2)
        root.left = TreeNode(2)
        root.right = TreeNode(5)
        root.right.left = TreeNode(5)
        root.right.right = TreeNode(7)
        
        sol = Solution()
        self.assertEqual(sol.findSecondMinimumValue(root), 5)

    def test_example_2_all_same_values(self):
        """Test Example 2: Tree where all node values are identical, returning -1."""
        # Tree structure: [2, 2, 2]
        root = TreeNode(2)
        root.left = TreeNode(2)
        root.right = TreeNode(2)
        
        sol = Solution()
        self.assertEqual(sol.findSecondMinimumValue(root), -1)

    def test_single_node_tree(self):
        """Test boundary case: A tree consisting of only a single root node."""
        root = TreeNode(10)
        
        sol = Solution()
        self.assertEqual(sol.findSecondMinimumValue(root), -1)

    def test_maximum_constraints_values(self):
        """Test constraints: Large node values near the 2^31 - 1 limit."""
        max_val = 2**31 - 1
        # Root holds the minimum of its children (42)
        root = TreeNode(42)
        root.left = TreeNode(42)
        root.right = TreeNode(max_val)
        
        sol = Solution()
        self.assertEqual(sol.findSecondMinimumValue(root), max_val)

    def test_skewed_value_distribution_left_branch(self):
        """Test deep tree scenario where the second minimum is on the left side."""
        root = TreeNode(3)
        root.left = TreeNode(3)
        root.right = TreeNode(3)
        root.left.left = TreeNode(4)
        root.left.right = TreeNode(3)
        
        sol = Solution()
        self.assertEqual(sol.findSecondMinimumValue(root), 4)

    def test_multiple_identical_second_minimums(self):
        """Test tree where the second minimum value appears in multiple independent leaf nodes."""
        root = TreeNode(10)
        root.left = TreeNode(10)
        root.right = TreeNode(10)
        root.left.left = TreeNode(20)
        root.left.right = TreeNode(10)
        root.right.left = TreeNode(10)
        root.right.right = TreeNode(20)
        
        sol = Solution()
        self.assertEqual(sol.findSecondMinimumValue(root), 20)